UnityProject/Assets/Scripts/Chemistry/ReactionLib.cs

```diff
@@ -70,167 +70,156 @@ private static void JsonImportInitialization()
 	{
-		var originalAmount = area[compatibleChem];
-		RemoveReagents(area, reaction, compatibleChem);
-		AddResults(area, reaction, compatibleChem, originalAmount);
+		float reactionMultiplier = GetReactionMultiplier(reagents, reaction); // Get the multiplier to multiply the reaction by
+		RemoveReagents(reagents, reaction, reactionMultiplier);
+		AddResults(reagents, reaction, reactionMultiplier);
 	}
 
-	private static void AddResults(Dictionary<string, float> area, Reaction reaction, string compatibleChem, float originalAmount)
+	private static void AddResults(Dictionary<string, float> reagents, Reaction reaction, float reactionMultiplier)
 	{
-		foreach (string chemical in reaction.Results.Keys)
+		foreach (var reagent in reaction.Results)
 		{
-			if (!area.ContainsKey(chemical)) { area[chemical] = 0; } //if result of the reaction doesn't already exist initialize it
-
-			area[chemical] += reaction.Results[chemical] * originalAmount / reaction.ReagentsAndRatio[compatibleChem]; //then adds the result
+			if (!reagents.ContainsKey(reagent.Key)) //if result of the reaction doesn't already exist initialize it
+			{
+				reagents[reagent.Key] = 0;
+			}
+			reagents[reagent.Key] += reagent.Value * reactionMultiplier; //then add the result adjusted by the leading reagent
 		}
 	}
 
 	/// <summary>
 	/// Does some mathematics to work out how much of each element to take away
 	/// </summary>
-	private static void RemoveReagents(Dictionary<string, float> area, Reaction reaction, string compatibleChem)
+	private static void RemoveReagents(Dictionary<string, float> reagents, Reaction reaction, float reactionMultiplier)
 	{
-		var originalAmount = area[compatibleChem];
-		var reAndRa = reaction.ReagentsAndRatio;
-
-		foreach (string chemical in reaction.ReagentsAndRatio.Keys)
+		foreach (var reagent in reaction.ReagentsAndRatio)
 		{
-			if (reaction.Catalysts.ContainsKey(chemical)) { continue; }
-
-			area[chemical] -= originalAmount * SwapFix(reAndRa[compatibleChem], reAndRa[chemical]);
+			if (reaction.Catalysts.ContainsKey(reagent.Key))
+			{
+				continue;
+			}
+			reagents[reagent.Key] -= reagent.Value * reactionMultiplier; // remove reagents adjusted by the leading reagent
 		}
 	}
 
-	/// <summary>
-	/// Finds the best chemical to do the reaction formula
-	/// </summary>
-	private static IEnumerable<string> CompatibleChemicals(Dictionary<string, float> area, Reaction reaction)
-	{
-		foreach (string chemical in reaction.ReagentsAndRatio.Keys)
-		{
-			if (!ReactionCompatible(area, reaction, chemical)) { continue; }
-
-			yield return chemical;
-		}
-	}
 
 	/// <summary>
-	/// Removes all chemicals from <paramref name="area"/> where the value is not more than zero
+	/// Gets a reaction that is possible with <paramref name="reagents"/>
 	/// </summary>
-	private static void RemoveEmptyChemicals(Dictionary<string, float> area)
+	private static Reaction GetValidReaction(Dictionary<string, float> reagents, float temperature)
```

Comment:
I had a talk in the discord and due to how this is programmed, If you have a tiny amount of catalysts  it could cause a lot of looping in the 
"while ((reaction = GetValidReaction(modifiedReagents, temperature)) != null)"
so a good suggestion could be to Have a minimum amount of catalysts before the reaction triggers?

Reply:
Should i maybe make it so that the catalyst cannot lead the reaction and it will just check if its present and get the leading reagent out of the non-catalyst ones? As it's not going to be used up, so there is no reason for it to lead. What do you think?

Reply:
I'll set it to the minimum in the recipe for now and we can change it later on